core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/EntityComponent.scala

```diff
@@ -352,21 +354,89 @@ trait EntityComponent {
+            } yield attributeRec
+          }
+
+          //actually build the records we're going to send to the db
+          val insertRecs = attributeMapToRecs(insertAttrs)
+          val updateRecs = attributeMapToRecs(updateAttrs)
+          val deleteIds = (for {
+            attributeName <- deletes
+          } yield existingAttrsToRecordIds.get(attributeName)).flatten
+
+          entityAttributeQuery.deltaAttrsAction(insertRecs, updateRecs, deleteIds, entityAttributeScratchQuery.insertScratchAttributes)
+        }
+      }
+    }
+
+    //save the deltas for some attributes on an existing entity. a little more database efficient than a "full" save, but requires the entity already exist.
+    def saveEntityDeltas(workspaceContext: SlickWorkspaceContext, entityRef: AttributeEntityReference, upserts: AttributeMap, deletes: Traversable[AttributeName]) = {
+      val deleteIntersectUpsert = deletes.toSet intersect upserts.keySet
+      if (upserts.isEmpty && deletes.isEmpty) {
+        DBIO.successful(()) //no-op
+      } else if (deleteIntersectUpsert.nonEmpty) {
+        DBIO.failed(new RawlsException(s"Can't saveEntityDeltas on $entityRef because upserts and deletes share attributes $deleteIntersectUpsert"))
+      } else {
+        getEntityRecords(workspaceContext.workspaceId, Set(entityRef)) flatMap { eRecs =>
+          if (eRecs.length != 1) {
+            throw new RawlsException(s"saveEntityDeltas looked up $entityRef expecting 1 record, got ${eRecs.length} instead")
+          }
+
+          val eRec = eRecs.head
+
+          upserts.keys.foreach { attrName =>
+            validateUserDefinedString(attrName.name)
+            validateAttributeName(attrName, eRec.entityType)
+          }
+
+          for {
+            _ <- applyAttributeDeltas(workspaceContext, eRec, upserts, deletes)
+            _ <- optimisticLockUpdate(eRec)
+          } yield {}
+        }
+      }
+    }
+
     private def lookupNotYetLoadedReferences(workspaceContext: SlickWorkspaceContext, entities: Traversable[Entity], alreadyLoadedEntityRecs: Seq[EntityRecord]): ReadAction[Seq[EntityRecord]] = {
-      val notYetLoadedEntityRecs = (for {
+      val allRefAttributes = (for {
         entity <- entities
         (_, attribute) <- entity.attributes
         ref <- attribute match {
           case AttributeEntityReferenceList(l) => l
           case r: AttributeEntityReference => Seq(r)
           case _ => Seq.empty
         }
-      } yield ref).toSet -- alreadyLoadedEntityRecs.map(_.toReference)
+      } yield ref).toSet
+
+      lookupNotYetLoadedReferences(workspaceContext, allRefAttributes, alreadyLoadedEntityRecs)
+    }
+
+    private def lookupNotYetLoadedReferences(workspaceContext: SlickWorkspaceContext, attrReferences: Set[AttributeEntityReference], alreadyLoadedEntityRecs: Seq[EntityRecord]): ReadAction[Seq[EntityRecord]] = {
```

Comment:
a bit of refactoring to provide "look up all attributes on these entities" and "look up all these references"